guides/builder.md

```diff
@@ -0,0 +1,274 @@
+
+The `remaining_to` field continues with the same behaviour as above.
+
+### Example
+With the metadata type above in mind, let's build a simple numscript.
+```elixir
+# metadata:
+%{
+  splits: [
+    %{
+      type: :percent,
+      amount: 20,
+      account: "some:destination"
+    }
+  ]
+  remaining_to: "another:destination",
+  percent_asset: "USD"
+}
+```
+Wich will generate the following numscript:
+```
+send [USD *] (
+  source = @user
+  destination = {
+    20% to @some:destination
+    remaining to @another:destination
+  }
+)
+```
+If the `:remaining_to` field were not defined, the remaining would be kept:
+```
+send [USD *] (
+  source = @user
+  destination = {
+    20% to @some:destination
+    remaining kept
+  }
+)
+```
+
+P.S.: The percent amount types will always be executed first, so we can assure that
+all of the percentages are based off of the initial value. This feature is still in
+development as we speak, but soon we will have more flexibility in those regards.
+
+## Examples
+### Example 1
+Both fixed and percent types:
+```elixir
+%{
+  splits: [
+    %{
+      type: :fixed,
+      amount: 5,
+      asset: "USD",
+      account: "some:destination:a"
+    },
+    %{
+      type: :percent,
+      amount: 20,
+      account: "some:destination:b"
+    }
+  ],
+  remaining_to: "some:destination:c",
+  percent_asset: "USD"
+}
+```
+Will generate:
+```
+send [USD *] (
+  source = @user
+  destination = {
+    20% to @some:destination:b
+    remaining kept
+  }
+)
+
+send [USD 500] (
+  source = @user
+  destination = @some:destination:a
+)
+```
+
+### Example 2
+Nested percent type plus fixed
+```elixir
+%{
+  splits: [
+    %{
+      type: :fixed,
+      amount: 12.5,
+      asset: "USD",
+      account: "some:destination:a"
+    },
+    %{
+      type: :percent,
+      amount: 30,
+      remaining_to: "remaining:destination:b",
+      splits: [
+        %{
+          type: :percent,
+          amount: 40,
+          account: "some:destination:c"
+        },
+        %{
+          type: :percent,
+          amount: 20,
+          account: "some:destination:b"
+        }
+      ]
+    }
+  ],
+  remaining_to: "remaining:destination:a",
+  percent_asset: "USD"
+}
+```
+Will generate:
+```
+send [USD *] (
+  source = @user
+  destination = {
+    30% to {
+      40% to @some:destination:c
+      20% to @some:destination:b
+      remaining to remaining:destination:b
+    }
+    remaining to remaining:destination:a
+  }
+)
+
+send [USD 1250] (
+  source = @user
+  destination = @some:destination:a
+)
+```
+Note that the `asset` field inside of nested destinations will be ignored, given that
+they all go under the same `send` scope.
+
+### Example 3
+Nested percent type plus fixed, and with fixed inside nests
+```elixir
+%{
+  splits: [
+    %{
+      type: :fixed,
+      amount: 10.5,
+      asset: "EUR",
+      account: "some:destination:a"
+    },
```

Comment:
Fiquei pensando: e se for importante para essa moeda ter mais de 2 decimais? Como podemos especificar que usa 3? ou 4?

Reply:
acho que dá pro campo `asset` ser um map em vez de uma string, por ex:

```elixir
    %{
      type: :fixed,
      amount: 1050,
      asset: %{name: "EUR", decimal_places: 2}
      account: "some:destination:a"
    }
```

e depois concatenar os dois
```elixir
"#{asset.name}/#{asset.decimal_places}
```

e acho que vou fazer uma struct, ta ficando muito grande e meio confuso

Reply:
Como é um valor externo (para ser usado com JSON) acho que não ajudaria muito ter um struct.

Seria bom se `decimal_places` for opcional, ou se usar só string (exemplo: "BRL") o padrão seja `2`, para ter a oportunidade de enviar o valor específico mas também não precisar definir sempre.

Reply:
Acho que só usar string facilita pra quem tá usando, e fazer um parse nessa string pra ver quantas casas decimais (ou não, se não tiver) não é nada muito complexo pra fazer do nosso lado.

E dois pontos importantes aqui:

1. Da pra colocar literalmente qualquer coisa nesse asset porque o Numscript não faz um parse pra ver se é um asset válido ou não
2. O comportamento do script é o mesmo independe das casas decimais e/ou do asset em si

Isso das casas decimais seria mais pra algo que a gente quisesse fazer internamente, como a feature de mostrar o valor decimal nas sctructs `Numscriptex.Posting` e `Numscriptex.Balance`